# Práctica 1: Introducción a la Neurona Artificial Simplificada (`neuron1.py`)

## 1. Introducción Teórica y Objetivos
En este cuaderno se implementa una neurona artificial básica de entrada única $x$ y un solo peso asociativo $w$, cuyo modelo matemático responde a la relación de salida lineal[cite: 7, 8]:
$$y = w \cdot x$$

El objetivo principal es lograr que la neurona aprenda la relación implícita presente en los datos de entrenamiento mediante la actualización progresiva del peso $w$ a lo largo de 30 épocas[cite: 7, 8].

### Datos de Entrenamiento
* **Entradas ($x$):** $[1, 2, 3, 4]$[cite: 7, 10]
* **Salidas Deseadas / Targets ($t$):** $[2, 4, 6, 8]$[cite: 7, 10]
* **Relación Matemática Implícita:** $t = 2x$[cite: 7]

### Mecanismo de Aprendizaje
El error individual de cada muestra se define como[cite: 14]:
$$e = t - p$$

El error promedio de la época ($cost$) se obtiene sumando los errores individuales y dividiendo entre la cantidad de muestras $n$[cite: 15]:
$$cost = \frac{\sum_{i=1}^{n} e_i}{n}$$

La regla de actualización directa del peso en cada época utiliza la tasa de aprendizaje $\eta$ mediante la fórmula[cite: 11, 16]:
$$w_{\text{nuevo}} = w_{\text{anterior}} + \eta \cdot cost$$

In [3]:
import time

inputs = [1, 2, 3, 4]
targets = [2, 4, 6, 8]

w = 0.1
learning_rate = 0.1
epochs = 30
pause_time = 0

def predict(i):
    return w * i

for epoch in range(epochs):
    predictions = [predict(i) for i in inputs]
    errors = [t - p for p, t in zip(predictions, targets)]
    cost = sum(errors) / len(targets)
    
    print("\n" + "=" * 60)
    print(f"ÉPOCA {epoch + 1}")
    print("=" * 60)
    for i, t, p, e in zip(inputs, targets, predictions, errors):
        print(f"Entrada: {i:>2} | Target: {t:>2} | Predicción: {p:>8.4f} | Error: {e:>8.4f}")
    
    print("-" * 60)
    print(f"Peso antes de actualizar: {w:.10f}")
    print(f"Error promedio: {cost:.6f}")
    
    w += learning_rate * cost
    
    print(f"Peso después de actualizar: {w:.10f}")
    time.sleep(pause_time)

test_inputs = [5, 6]
test_targets = [10, 12]

print("\n" + "=" * 60)
print("PRUEBA FINAL")
print("=" * 60)

test_predictions = [predict(i) for i in test_inputs]

for i, t, p in zip(test_inputs, test_targets, test_predictions):
    print(f"Entrada: {i:>2} | Target: {t:>2} | Predicción: {p:>8.4f}")


ÉPOCA 1
Entrada:  1 | Target:  2 | Predicción:   0.1000 | Error:   1.9000
Entrada:  2 | Target:  4 | Predicción:   0.2000 | Error:   3.8000
Entrada:  3 | Target:  6 | Predicción:   0.3000 | Error:   5.7000
Entrada:  4 | Target:  8 | Predicción:   0.4000 | Error:   7.6000
------------------------------------------------------------
Peso antes de actualizar: 0.1000000000
Error promedio: 4.750000
Peso después de actualizar: 0.5750000000

ÉPOCA 2
Entrada:  1 | Target:  2 | Predicción:   0.5750 | Error:   1.4250
Entrada:  2 | Target:  4 | Predicción:   1.1500 | Error:   2.8500
Entrada:  3 | Target:  6 | Predicción:   1.7250 | Error:   4.2750
Entrada:  4 | Target:  8 | Predicción:   2.3000 | Error:   5.7000
------------------------------------------------------------
Peso antes de actualizar: 0.5750000000
Error promedio: 3.562500
Peso después de actualizar: 0.9312500000

ÉPOCA 3
Entrada:  1 | Target:  2 | Predicción:   0.9313 | Error:   1.0687
Entrada:  2 | Target:  4 | Predicción:   1.8625

## 2. Interpretación de Resultados
* **Convergencia del Peso:** El peso inicial $w = 0.1$ incrementa progresivamente hasta aproximarse a $w \approx 2.0$[cite: 10, 11, 17].
* **Disminución del Error:** El error promedio pasa de $4.75$ en la primera época a valores prácticamente nulos[cite: 11, 15].
* **Capacidad de Generalización:** En la prueba final con $x=5$ y $x=6$, la neurona genera predicciones muy cercanas a $10.0$ y $12.0$, demostrando que aprendió la función $y = 2x$[cite: 17, 18].